controller/src/main/java/io/pravega/controller/server/bucket/BucketManagerLeader.java

```diff
@@ -0,0 +1,201 @@
+                    controllerChange.release();
+                    break;
+                case ERROR:
+                    // This event should be due to ZK connection errors and would have been received by the monitor too,
+                    // hence not handling it explicitly here.
+                    log.error("{}: Received error event when monitoring the pravega host cluster, ignoring...", serviceType);
+                    break;
+            }
+        });
+
+        // Keep looping here as long as possible to stay as the leader and exclusively monitor the pravega cluster.
+        while (true) {
+            try {
+                if (suspended.get()) {
+                    log.info("{}: Monitor is suspended, waiting for notification to resume.", serviceType);
+                    suspendMonitor.await();
+                    log.info("{}: Resuming monitor.", serviceType);
+                }
+
+                controllerChange.await();
+                log.info("{}: Received distribute buckets.", serviceType);
+
+                // Wait here until distribution can be performed.
+                waitForReDistribute();
+
+                // Clear all events that has been received until this point since this will be included in the current
+                // distribution operation.
+                controllerChange.reset();
+                triggerDistribution(pravegaServiceCluster);
+            } catch (InterruptedException e) {
+                log.warn("{}: Leadership interrupted, releasing monitor thread.", serviceType);
+                Thread.currentThread().interrupt();
+                // Stop watching the pravega cluster.
+                pravegaServiceCluster.close();
+
+                throw e;
+            } catch (Exception e) {
+                // We will not release leadership if in suspended mode.
+                if (!suspended.get()) {
+                    log.warn("{}: Failed to perform distribution, relinquishing leadership.", serviceType);
+                    // Stop watching the pravega cluster.
+                    pravegaServiceCluster.close();
+                    throw e;
+                }
+            }
+        }
+
+    }
+
+    /**
+     * Blocks until the redistribute interval. This wait serves multiple purposes:
+     * -- Ensure redistribution does not happen in quick succession since its a costly cluster level operation.
+     * -- Clubs multiple host events into one to reduce redistribution operations. For example:
+     *      Fresh cluster start, cluster/multi-host/host restarts, etc.
+     */
+    private void waitForReDistribute() throws InterruptedException, ExecutionException {
+        log.info("{}: Waiting for {} seconds before attempting to distribute.", serviceType,
+                minBucketRedistributionIntervalInSeconds.getSeconds());
+        Futures.delayedFuture(minBucketRedistributionIntervalInSeconds, executorService).get();
```

Comment:
This line is just `Thread.sleep` with extra steps, please simplify if the future is not actually going to be used as a future.